### ISSUE TITLE
Is there any roadmap for Bosque?

### ISSUE THREAD

**Issue by catchmareck** [microsoft/BosqueLanguage#365]
I'm wondering if there is any development roadmap available where we can see what are the future plans for Bosque. I'm thinking about the bigger picture here, like future integrations, ideas, features, etc.

Basically, where this project is heading - is there an intention to make it a language used in many solutions in practice (like TS or C#), or is it rather planned to keep it as an experiment and never encourage people to use it on production? (I know it's not easy to make people use some language but I'm just curious if that is your intention or you have some other plans in mind. Do you know what I mean?)

Is there any place where we can see it? Or could you please share your thoughts about it?

**Comment by catchmareck** [comment]
@mrkmarron I just noticed that you have added news about Bosque v1 and the branch we can work on. From the git diff, I can tell there are a lot of changes - new features mostly. The question is: how can we contribute? Are there any specific aspects that have a higher priority than others?

Also, is there any full changelog of what's changed? I guess the brief description in the Readme does not cover all of the changes.

Basically, the main question from the original post still stands.

Thanks for sharing your work publicly so we can learn from it, test things, and contribute to the Bosque project! :)

**Comment by mrkmarron** [comment]
Hi @catchmareck thanks for the comments. Yes there are a large number of changes there and more are coming!

My major goal at this point is to get a version of the "road_to_1_0" repository in a state where you can run the type checker (and a new verifier module). I expect this to happen right after the holidays and a new compiler backend should land shortly after that. 

Once this happens there will be a fair number of shallow parser/type-checker bugs that will need to be found/fixed so any bug reports and/or fixes will be incredibly helpful. 

A couple of interesting items that are a bit more involved but that are on the fun side are:
1. An [inliner](https://en.wikipedia.org/wiki/Inline_expansion) at the IR level. For languages like Bosque this is a basic and very effective compiler optimization. It should also help a lot with how we do lambda instantiation. It might also be interesting to support code-deduplication as well. 
2. A [Common Subexpression Elimination](https://en.wikipedia.org/wiki/Common_subexpression_elimination) pass (perhaps on the SSA IR form) would be another very useful optimization. And since Bosque is fully functional/immutable at the IR level it is possible to be highly aggressive here. So, it would be interesting (perhaps even worth writing-up) to see how much more effective.
3. A more free-form item is integrating with Node.js. A large number of changes in this version are to get all the parts of "API Types" to line-up cleanly. Roughly this means that we can give a type signature in Bosque that will map to an easily JSON parsable object. It would be fun to have a pipeline that takes a set of Bosque files and entrypoints, and builds a NAPI module that exposes the endpoints using API Types so they can be called directly from JavaScript.

If there is interest in any of these I am happy to spend a bit of time mentoring.

The change-log is going to take some time to writeup and I want to update all the other documentation as well so it may be a bit before that all happens.

I am glad to hear that you have found the project interesting/useful. It has been a challenge to open-source this from the beginning as we experiment and struggle with things publicly. I know it can also be challenging for folks in the community as they try and engage with things. So, thanks for your (and everyone's) contributions and for your encouragement.

**Comment by catchmareck** [comment]
Hi @mrkmarron, sorry for the lack of response - we all are always busy these days ;(

Just wanted to know if any major changes are going to happen in the near future?

Also, I'm curious about point 3. (integrating with Node.js). I'm passionate about Node and would love to know more about this feature in more detail! Could you please elaborate on this one?

**Comment by mrkmarron** [comment]
Hi, yes some big changes should be landing soon! 

One aspect that I have been working on, that is currently under-documented, is APITypes. These types have several nice properties:
- They are entirely defined in terms of primitive or self describing types -- Int, Bool, String, Tuples (of APITypes), Records (of APITypes), List (of APITypes), etc.
- They map very nicely to the type of things that are JSON objects
- They can be used to automatically generate parser/validator logic for parsing JSON strings or objects

So, idea is to take an interface that uses APITypes, generate the appropriate object validator code, and then generate JS bindings to call this. One option I like is to use [NAPI](https://nodejs.org/api/n-api.html) to do a fast native implementation another, since it looks like the native compiler will be delayed a bit, is to revive the interpreter (which was in JS) and just run the bundle that way. There is some overhead bumping between real JS and the Bosque implementation but it should allow us to use the rich Node ecosystem and then plug in Bosque code implementation where they make sense.
